---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
# Correlation Tree Analysis

<!-- badges: start -->
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/ba04cd22d16047bb831608b9a7a6702f)](https://www.codacy.com/app/abichat/correlationtree_analysis?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=abichat/correlationtree_analysis&amp;utm_campaign=Badge_Grade)
![Last-changedate](https://img.shields.io/badge/last%20change-`r gsub('-', '--', Sys.Date())`-yellowgreen.svg)
[![Journal](https://img.shields.io/badge/published-bioRxiv-blue)](https://www.biorxiv.org/content/10.1101/2020.01.31.928309v1)
<!-- badges: end -->

This repository contains analysis done in the paper **Incorporating phylogenetic information in microbiome abundance studies has no effect on detection power and FDR control** ([bioRxiv](https://www.biorxiv.org/content/10.1101/2020.01.31.928309v1)).

Some results might be slightly different from those in the article due to seed choice or limited number of replications in simulations.


## Structure of the repository

### Forest

Each folder is named after a dataset and contains several files:

- `anaylsis_dataset.R`: R script that performs the analysis, 
- `dataset.biom`: biom-format file with the abundance if used,
- `phytree_dataset.nwk`: phylogeny or taxonomy in the newick format if used,
- `XXX.png`: figure. 

Each of these scripts:

1. loads the data,
2. generates the forest of trees,
3. computes pairwise distances between trees (with BHV and RF),
4. performs PCoAs (one per ditance),
5. draws individual plots.

Figures 3, S1 and S2 are drawn with these scripts.

### Real Datasets

This folder contains scripts to do differentialy abundance studies on datasets Chaillou, Chlamydiae and Zeller (genus and MSP level).

Figures 6, 7, S5, S6, S7 and S8 are drawn with these scripts.


### Simulations

This folder contains scripts that simulates datasets according to parametric and non parametric schemes.

Figures 4, 5, S3 and S4 are drawn with these scripts.

### Figures

This folder contains scripts that draw every figure in the article. Each script takes its input in the corresponding folders of the repository.


## Package versions used for the analysis

```{r used_packages, echo=FALSE, message=FALSE, cacha=TRUE, cache.lazy=FALSE}
library(tidyverse)

all_R <- list.files(path = ".", full.names = TRUE, pattern = ".*\\.R$", recursive = TRUE)

packages <-
  all_R %>%
  map(parse) %>%
  map(as.character) %>%
  unlist() %>%
  str_extract_all("^library(.*)$") %>%
  unlist() %>%
  str_remove_all("library\\(|\\)")

df_package_version <-
  packages %>%
  c(., "dplyr", "tidyr", "ggplot2", "purrr", "readr", "stringr", "forcats", "furrr") %>%
  unique() %>%
  sort() %>%
  enframe(value = "Package", name = NULL) %>%
  mutate(Version = map_chr(Package, purrr::compose(as.character, packageVersion)))

knitr::kable(df_package_version)
```

To install non-CRAN packages, run these lines:
```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("biomformat")
BiocManager::install("curatedMetagenomicData")
BiocManager::install("ggtree")
BiocManager::install("phyloseq")
BiocManager::install("multtest") # dependancy for structSSI

if (!requireNamespace("remotes", quietly = TRUE))
    install.packages("remotes")
remotes::install_github("abichat/correlationtree")
remotes::install_github("abichat/evabic")
remotes::install_url("https://cran.r-project.org/src/contrib/Archive/structSSI/structSSI_1.1.1.tar.gz") # archived from CRAN
```

