---
output: github_document
bibliography: datasets.bib
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
# Correlation Tree Analysis

<!-- badges: start -->
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/ba04cd22d16047bb831608b9a7a6702f)](https://www.codacy.com/app/abichat/correlationtree_analysis?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=abichat/correlationtree_analysis&amp;utm_campaign=Badge_Grade)
![Last-changedate](https://img.shields.io/badge/last%20change-`r gsub('-', '--', Sys.Date())`-yellowgreen.svg)
[![Journal](https://img.shields.io/badge/published-bioRxiv-blue)](https://www.biorxiv.org/content/10.1101/2020.01.31.928309v1)
<!-- badges: end -->

This repository contains analysis done in the paper **Incorporating phylogenetic information in microbiome abundance studies has no effect on detection power and FDR control** ([bioRxiv](https://www.biorxiv.org/content/10.1101/2020.01.31.928309v1)).

Some results might be slightly different from those in the article due to seed choice or limited number of replications in simulations.


## Structure of the repository

### Forest

Each subfolder is named after the studied dataset and contains several files:

- a script that performs the analysis (`analysis_dataset.R`),
- several `.png` figures corresponding to different visualizations and distances,
- saved intermediary results to avoid long recomputation (`.rds`),
- evetually biom (`.biom`) or newick (`.nwk`) files for input data such as abundnace tables or phylogeny.

Each R script:

1. loads the data,
2. generates the forest of trees,
3. computes pairwise distances between trees (with BHV and RF),
4. performs PCoAs (one per distance),
5. draws individual plots:
    1. distance from each tree to the correlation tree,
    2. projection of the forest on the two first axis.

### Real Datasets

This folder contains scripts to do differentialy abundance studies on datasets Chaillou, Chlamydiae and Zeller (genus and MSP level).

As previously, it contains the R script, plots (`.png`), intermediary results (`.rds`) and eventually biological input data (`.biom` and `.nwk`).

Each script compares correlation and taxonomy (or phylogeny) in term of detected species. It uses hierarchical FDR (from `structSSI`) for both Chaillou and Chlamydiae datasets, and z-scores smoothing (from `StructFDR`) for datasets from Zeller.

### Simulations

This folder contains scripts that simulates datasets according to parametric (P) and non parametric (NP) schemes.

The parametric simulation mimics the scheme used in @xiao_false_2017. It fits a negative-binomial Dirichlet-Multinomial (DM) distibution on Wu dataset and generates new differentially abundance datasets.

The non-parametric simulation uses a real dataset from @brito_mobile_2016. It generates differentially abundant species by applying specified fold-change in half of the samples.

For each set of parameters, at least 600 replication were done. As this is really time consumming, preprocessed data are saved in `.rds` files.


### Figures

This folder contains the scripts used to produce every figure in the article. Each script takes its input in the folders of the repository and is named after the Figure number in the article (`Figure_1.R`, `Figure_S1.R`, etc). 

## Reproducibility and packages

Analysis ran under R version `r paste(R.version$major, R.version$minor, sep = ".")`.

```{r used_packages, echo=FALSE, message=FALSE, cacha=TRUE, cache.lazy=FALSE}
library(tidyverse)

all_R <- list.files(path = ".", full.names = TRUE, pattern = ".*\\.R$", recursive = TRUE)

packages <-
  all_R %>%
  map(parse) %>%
  map(as.character) %>%
  unlist() %>%
  str_extract_all("^library(.*)$") %>%
  unlist() %>%
  str_remove_all("library\\(|\\)")

df_package_version <-
  packages %>%
  c(., "dplyr", "tidyr", "ggplot2", "purrr", "readr", "stringr", "forcats", "furrr") %>%
  unique() %>%
  sort() %>%
  enframe(value = "Package", name = NULL) %>%
  mutate(Version = map_chr(Package, purrr::compose(as.character, packageVersion)))

knitr::kable(df_package_version)
```

To install non-CRAN packages, run these lines:
```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("biomformat")
BiocManager::install("curatedMetagenomicData")
BiocManager::install("ggtree")
BiocManager::install("phyloseq")
BiocManager::install("multtest") # dependency for structSSI

if (!requireNamespace("remotes", quietly = TRUE))
    install.packages("remotes")
remotes::install_github("abichat/correlationtree")
remotes::install_github("abichat/evabic")
remotes::install_url("https://cran.r-project.org/src/contrib/Archive/structSSI/structSSI_1.1.1.tar.gz") # archived from CRAN
```

## Datasets


| Dataset    | Biome   | Rank  | Taxa   | Samples | Analysis                                                    | Publication            |
|------------|---------|-------|--------|---------|-------------------------------------------------------------|------------------------|
| Chlamydiae | Varied  | OTU   | 21     | 26      | [Tree](forests/chlamydiae) & [DA](real_datasets/chlamydiae) | @caporaso_global_2011  |  
| Ravel      | Vaginal | Genus | 40     | 396     | [Tree](forests/ravel)                                       | @ravel_vaginal_2011    |
| Wu         | Gut     | OTU   | 400    | 98      | [Simulations](simulations/parametric)                       | @wu_linking_2011       |
| Zeller     | Gut     | Genus | 119    | 199     | [Tree](forests/zeller/) & [DA](real_datasets/zeller_genus)  | @zeller_potential_2014 | 
| Zeller MSP | Gut     | MSP   | 878    | 199     | [DA](real_datasets/zeller_msp)                              | @zeller_potential_2014 |
| Chaillou   | Food    | OTU   | 499/97 | 64      | [Tree](forests/chaillou) & [DA](real_datasets/chaillou)     | @chaillou_origin_2015  | 
| Brito      | Gut     | OTU   | 77     | 112     | [Simulations](simulations/non_parametric)                   | @brito_mobile_2016     |


## Bibliography
